File: Lua/bench.lua
-- require "benchmark"

-- GC.disable

-- time = Benchmark.realtime do
    
num_rows = 100000
num_cols = 10
count = 0
xx = string.rep("x", 1000)

function shell(command)
    local handle = io.popen(command)
    local result = handle:read("*a")
    handle:close()
    return result
end

local unistd = require 'posix.unistd'

function rss()
    print(tonumber(shell("ps -o rss= -p " .. unistd.getpid())) // 1024 .. " MB")
end

function gen()
    count = count+1
    return count .. xx
end

function init(size, generator)
    local array = {}
    for i=1,size do
        array[i] = generator()
    end
   
    return array
end

function map(array, func)
    local new_array = {}
    for i,v in ipairs(array) do
      new_array[i] = func(v)
    end
    return new_array
end

function makeCSV()
    local data = init(num_rows, function (row) return init(num_cols, gen) end)
    print(count)
    local csv = table.concat(map(data, function (row) return table.concat(row, ",") end), "\n")
    rss()
    return csv
end
    
for i=1,10 do
    print("iteration " .. i)
    print(#(makeCSV()))
end

-- print time.round(2)

File: C++/bench.cpp
// bench.cpp

#include <iostream>
#include <vector>
#include <sstream>
#include <ctime>
#include <unistd.h>

const int num_rows = 100000;
const int num_cols = 10;
int count = 0;
std::string xx(1000, 'x');

void printMemoryUsage() {
    std::cout << (static_cast<int>(system(("ps -o rss= -p " + std::to_string(getpid())).c_str())) / 1024) << " MB" << std::endl;
}

std::string gen() {
    count++;
    return std::to_string(count) + xx;
}

std::string makeCSV() {
    std::vector<std::vector<std::string>> data(num_rows, std::vector<std::string>(num_cols));
    
    for (int i = 0; i < num_rows; ++i) {
        for (int j = 0; j < num_cols; ++j) {
            data[i][j] = gen();
        }
    }
    
    std::cout << count << std::endl;

    std::ostringstream csv;
    for (const auto& row : data) {
        csv << row[0];
        for (int j = 1; j < num_cols; ++j) {
            csv << "," << row[j];
        }
        csv << "\n";
    }

    printMemoryUsage();
    return csv.str();
}

int main() {
    clock_t start_time = clock();

    printMemoryUsage();
    for (int i = 0; i < 10; ++i) {
        std::cout << makeCSV().length() << std::endl;
    }

    clock_t end_time = clock();
    double time = static_cast<double>(end_time - start_time) / CLOCKS_PER_SEC;
    std::cout << "Time: " << time << " seconds" << std::endl;

    return 0;
}
